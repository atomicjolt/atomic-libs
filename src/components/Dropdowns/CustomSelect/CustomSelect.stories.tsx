import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import { InputControls } from "../../storybook";
import { CustomSelect } from ".";
import { Item, Section } from "../../Collection";

const meta: Meta<typeof CustomSelect> = {
  title: "Dropdowns/Selection/CustomSelect",
  component: CustomSelect,
  parameters: {
    layout: "centered",
  },
  argTypes: {
    ...InputControls,
    selectedKey: {
      control: "select",
      description:
        "Control the value of the selected key in a controlled component",
      options: ["val1", "val2", "val3", null],
    },
    defaultSelectedKey: {
      control: "false",
      description:
        "Control the value of the selected key in an uncontrolled component",
    },
    children: {
      control: false,
    },
    isOpen: {
      control: "boolean",
      description: "Control the open state of the menu",
    },
    defaultOpen: {
      control: "boolean",
      description: "Control the inital state of the menu",
    },
    onOpenChange: {
      control: false,
      action: "onOpenChange",
      description: "Callback for when the menu is opened or closed",
      table: {
        category: "Events",
      },
    },
    onSelectionChange: {
      action: "onSelectionChange",
      control: false,
      description: "Callback for when the selection changes",
      table: {
        category: "Events",
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof CustomSelect>;

export const Primary: Story = {
  args: {
    size: "medium",
    menuSize: "medium",
    variant: "default",
    label: "Custom Select Label",
    children: [
      <Item key="val1">Item 1</Item>,
      <Item key="val2">Item 2</Item>,
      <Item key="val3">Item 3</Item>,
    ],
  },
};

export const WithSections: Story = {
  args: {
    ...Primary.args,
    children: [
      <Item key="val1">Item 1</Item>,
      <Item key="val2">Item 2</Item>,
      <Item key="val3">Item 3</Item>,
      <Section title="Section Title">
        <Item key="val4">Item 4</Item>
        <Item key="val5">Item 5</Item>
      </Section>,
    ],
  },
};

export const Floating: Story = {
  args: {
    ...Primary.args,
    defaultSelectedKey: "undefined",
    variant: "floating",
  },
};

export const Searchable: Story = {
  args: {
    ...Primary.args,
    isSearchable: true,
    searchPlaceholder: "Search",
  },
};
