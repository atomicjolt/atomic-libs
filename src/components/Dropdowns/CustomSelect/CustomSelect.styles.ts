import styled from "styled-components";
import mixins from "../../../styles/mixins";

export const SearchListItem = styled.li`
  border-bottom: 1px solid var(--option-border-clr);

  i {
    position: absolute;
    right: 8px;
    top: 50%;
    transform: translateY(-50%);
    font-size: 2.4rem;
    color: var(--neutral400);
  }
`;

export const SearchInput = styled.input`
  ${mixins.Regular}
  border: none;
  background-color: var(--option-bg-clr);
  height: var(--option-height);
  width: 100%;
  padding: 0 32px 0 var(--option-padding-horizontal);
  font-size: var(--option-font-size);
  color: var(--option-text-clr);

  &:focus {
    outline: 2px solid var(--outline-clr-primary);
    outline-offset: -2px;
  }
`;

export const ButtonText = styled.span`
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const SelectButtonWrapper = styled.div`
  position: relative;
`;
