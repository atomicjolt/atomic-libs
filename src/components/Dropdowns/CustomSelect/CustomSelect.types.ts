import React from "react";
import { InputProps } from "../../../types";

export interface CustomSelectProps<T>
  extends InputProps<T | null, Element, React.MouseEvent<Element>> {
  readonly children: React.ReactElement | React.ReactElement[];
}

export interface CustomSelectContext<T> {
  currentValue: T | null;
  onClick: (value: T, event: React.MouseEvent<Element>) => void;
}
