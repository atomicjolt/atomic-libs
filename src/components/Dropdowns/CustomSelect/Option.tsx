import React, { useContext } from "react";

export interface CustomSelectOptionProps<T> {
  value: T;
  children: React.ReactNode;
}

/** An Option in the `CustomSelect` */
export function Option<T>({ value, children }: CustomSelectOptionProps<T>) {
  return null;
}

/** A `CustomSelect` `Option` item with `null` as the value */
export function EmptyOption({ children }: { children?: React.ReactNode }) {
  return <Option value={null}>{children}</Option>;
}
