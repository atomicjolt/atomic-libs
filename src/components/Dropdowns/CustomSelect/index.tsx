import React, { useRef } from "react";
import { SelectProps, useSelectState } from "react-stately";
import { HiddenSelect, useSelect } from "react-aria";
import classNames from "classnames";
import {
  AriaProps,
  ExtendedSize,
  FieldBaseProps,
  HasVariant,
} from "../../../types";
import { DropdownWrapper } from "../Dropdowns.styles";
import { useVariantClass } from "../../../hooks";
import { ButtonText } from "./CustomSelect.styles";
import { FieldWrapper } from "../../Utility/FieldWrapper";
import Button from "../../Buttons/Button";
import MaterialIcon from "../../Icons/MaterialIcon";
import Popover from "../../Utility/Popover";
import { UnmanagedListBox } from "../../Utility/ListBox";

export type CustomSelectVariants = "default" | "floating";

export interface CustomSelectProps<T extends object>
  extends AriaProps<SelectProps<T>>,
    Omit<FieldBaseProps, "isReadOnly">,
    HasVariant<CustomSelectVariants> {
  /** The size of the menu. Defaults to `auto` */
  menuSize?: ExtendedSize;
}

/** A custom version of the builtin `select` component to
 * allow for consistent styling & an extended feature set */
export function CustomSelect<T extends object>(props: CustomSelectProps<T>) {
  const state = useSelectState<T>(props);
  const ref = useRef(null);
  const {
    labelProps,
    triggerProps,
    valueProps,
    menuProps,
    errorMessageProps,
    descriptionProps,
  } = useSelect(props, state, ref);

  const {
    className,
    label,
    hideLabel,
    error,
    message,
    isInvalid,
    isDisabled,
    isRequired,
    name,
    size = "medium",
    menuSize = "auto",
    placeholder = "Select an option",
    variant = "default",
  } = props;

  const variantClassName = useVariantClass("aje-dropdown", variant);

  return (
    <DropdownWrapper
      className={classNames("aje-dropdown", variantClassName, className, {
        "has-selection": state.selectedItem,
      })}
      error={isInvalid}
      disabled={isDisabled}
      required={isRequired}
      size={size}
    >
      <HiddenSelect
        isDisabled={isDisabled}
        state={state}
        triggerRef={ref}
        label={label}
        name={name}
      />

      <FieldWrapper
        label={label}
        labelProps={labelProps}
        message={message}
        messageProps={descriptionProps}
        error={error}
        errorProps={errorMessageProps}
        hideLabel={hideLabel}
        floating={variant === "floating"}
      >
        <Button
          {...triggerProps}
          ref={ref}
          variant="border"
          size={size}
          isDisabled={isDisabled}
        >
          <ButtonText {...valueProps}>
            {state.selectedItem
              ? state.selectedItem.rendered
              : variant === "default"
              ? placeholder
              : ""}
          </ButtonText>
          <MaterialIcon icon="arrow_drop_down" />
        </Button>
      </FieldWrapper>
      {state.isOpen && (
        <Popover state={state} triggerRef={ref} placement="bottom start">
          {/* @ts-ignore */}
          <UnmanagedListBox {...menuProps} state={state} size={menuSize} />
        </Popover>
      )}
    </DropdownWrapper>
  );
}
