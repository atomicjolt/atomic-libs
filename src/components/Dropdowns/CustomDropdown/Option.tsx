import React, { useContext } from "react";
import cn from "classnames";
import { Context } from "./utils";

interface CustomSelectOptionProps<T> {
  value: T;
  children: React.ReactNode;
}

/** An Option in the `CustomDropdown` */
export function Option<T>({ value, children }: CustomSelectOptionProps<T>) {
  const { currentValue, onClick } = useContext(Context);

  return (
    <li
      className={cn("aje-combobox__option", {
        "is-focused": value == currentValue,
      })}
      role="option"
      aria-selected={value == currentValue}
      onClick={(e) => onClick(value, e)}
    >
      {children}
    </li>
  );
}

/** A `CustomDropdown` `Option` item with `null` as the value */
export function EmptyOption({ children }: { children?: React.ReactNode }) {
  return <Option value={null}>{children}</Option>;
}
