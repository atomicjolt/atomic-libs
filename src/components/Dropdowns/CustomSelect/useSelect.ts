import React, { useRef, useState } from "react";

import { useBool, useClickOutside } from "../../../hooks";
import { EventHandler, FilterFunction } from "../../../types";
import { OptionProps } from "../Option";

// TYPES --------------------------------------------------------
type SelectOnChangeHandler<T> = EventHandler<
  T,
  React.MouseEvent<Element> | React.KeyboardEvent<Element>
>;

interface UseSelectOptions<T, ChildrenProps extends OptionProps<T>> {
  readonly children:
    | React.ReactElement<ChildrenProps>
    | React.ReactElement<ChildrenProps>[];
  readonly searchable: boolean;
  readonly filterOptions: FilterFunction<ChildrenProps>;
  readonly value: T | T[];
  readonly onChange?: SelectOnChangeHandler<T> | SelectOnChangeHandler<T[]>;
}

interface UseSelectMenu {
  readonly opened: boolean;
  readonly open: () => void;
  readonly close: () => void;
  readonly toggle: () => void;
}

interface UseSelectSearch {
  readonly value: string;
  readonly onChange: React.ChangeEventHandler<HTMLInputElement>;
}

interface UseSelectReturn<T, ChildrenProps extends OptionProps<T>> {
  readonly selectedValues: T[];
  readonly selectedOption: ChildrenProps;
  readonly selectedIndex: number;
  readonly multiselect: boolean;
  readonly ref: React.RefObject<any>;
  readonly options: ChildrenProps[];
  readonly menu: UseSelectMenu;
  readonly search: UseSelectSearch;
  readonly handleKeyPress: React.KeyboardEventHandler;
  readonly isSelected: (value: T) => boolean;
  readonly isFocused: (value: T) => boolean;
  readonly onChange: SelectOnChangeHandler<T>;
}

// Functions --------------------------------------------------------

function handleMultiSelect<T>(
  value: T | T[]
): [values: T[], multiselect: boolean] {
  let selectedValues: T[];
  let multiselect = false;
  if (Array.isArray(value)) {
    selectedValues = value;
    multiselect = true;
  } else {
    selectedValues = [value];
  }

  return [selectedValues, multiselect];
}

export default function useSelect<
  SelectedValue,
  ChildProps extends OptionProps<SelectedValue>
>(
  opts: UseSelectOptions<SelectedValue, ChildProps>
): UseSelectReturn<SelectedValue, ChildProps> {
  // SETUP -------------------------------------

  const { value, children, onChange, searchable, filterOptions } = opts;
  const [selectedValues, multiselect] = handleMultiSelect(value);

  let options: ChildProps[] = React.Children.map(
    children,
    (child) => child.props
  );

  const selectedIndex = options.findIndex((o) =>
    selectedValues.includes(o.value)
  );

  const selectedOption = options[selectedIndex];
  // HOOKS -------------------------------------

  const [search, setSearch] = useState<string>("");

  if (searchable && search) {
    options = filterOptions(search, options);
  }

  const ref = useRef<HTMLDivElement | null>(null);

  const [menuActive, toggleMenu, openMenu, closeMenu] = useBool(false);

  const [focused, setFocused] = useState(selectedIndex || 0);

  useClickOutside(
    ref,
    () => {
      if (menuActive) {
        handleClose(true);
      }
    },
    { enabled: menuActive }
  );

  // CALLBACKS -------------------------------------
  const isSelected = (value: SelectedValue) => selectedValues.includes(value);
  const isFocused = (value: SelectedValue) => value == options[focused]?.value;

  const handleClose = (force: boolean = false) => {
    if (!multiselect || force) {
      if (focused !== selectedIndex) {
        setFocused(selectedIndex);
      }
      closeMenu();
    }
  };

  const onChangeWrapper: SelectOnChangeHandler<SelectedValue> = (v, e) => {
    handleClose();
    if (!onChange) return;

    // TODO: figure out a better way to disciminate between these types
    // rather than just asserting it manually
    if (multiselect) {
      const multiOnChange = onChange as SelectOnChangeHandler<SelectedValue[]>;
      if (selectedValues.includes(v)) {
        multiOnChange(
          selectedValues.filter((s) => s !== v),
          e
        );
      } else {
        multiOnChange([...selectedValues, v], e);
      }
    } else {
      const singleOnChange = onChange as SelectOnChangeHandler<SelectedValue>;
      singleOnChange(v, e);
    }
  };

  const handleKeyPress: React.KeyboardEventHandler = (e) => {
    if (e.key === " ") {
      openMenu();
    } else if (e.key === "Escape") {
      if (menuActive) {
        handleClose(true);
      } else {
        ref.current?.blur();
      }
    } else if (menuActive) {
      switch (e.key) {
        case "ArrowDown":
          setFocused(Math.min(focused + 1, options.length - 1));
          break;
        case "ArrowUp":
          setFocused(Math.max(focused - 1, 0));
          break;
        case "Enter":
          onChangeWrapper(options[focused].value, e);
          handleClose();
        default:
          break;
      }
    }
  };

  const onSearch: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    setSearch(e.target.value);
  };

  const menu = {
    opened: menuActive,
    open: openMenu,
    close: () => handleClose(true),
    toggle: toggleMenu,
  };

  const searchData = {
    value: search,
    onChange: onSearch,
  };

  return {
    selectedValues,
    selectedOption,
    selectedIndex,
    multiselect,
    ref,
    options,
    handleKeyPress,
    isFocused,
    isSelected,
    onChange: onChangeWrapper,
    menu,
    search: searchData,
  };
}
