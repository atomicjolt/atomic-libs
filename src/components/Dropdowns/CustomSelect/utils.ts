import React, { useMemo } from "react";
import { CustomSelectContext } from "./CustomSelect.types";
import { EmptyOption, Option } from "../Option";

export const Context = React.createContext<CustomSelectContext<any>>(
  {} as CustomSelectContext<any>
);

export function useSelectedChild<T>(value: T, children: React.ReactNode) {
  return useMemo(() => {
    const childArray = React.Children.toArray(children) as React.ReactElement[];
    childArray.forEach((c) => {
      if (c.type !== Option && c.type !== EmptyOption) {
        throw new Error(
          "All children of CustomSelect must be Option or EmptyOption"
        );
      }
    });
    return childArray.find((c) => c.props["value"] == value);
  }, [value]);
}
