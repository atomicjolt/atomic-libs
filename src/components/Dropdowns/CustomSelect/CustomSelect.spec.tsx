import React from "react";
import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import CustomSelect from ".";
import Option from "../Option";

describe("matches snapshots", () => {
  const shared = {
    label: "Custom dropdown label",
    value: "opt1",
    onChange: () => {},
  };

  test("matches default variant", () => {
    const result = render(
      <CustomSelect {...shared}>
        <Option value="opt1">Option 1</Option>
        <Option value="opt2">Option 2</Option>
      </CustomSelect>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });

  test("matches multi-select enabled", () => {
    const result = render(
      <CustomSelect {...shared} value={["opt1", "opt2"]}>
        <Option value="opt1">Option 1</Option>
        <Option value="opt2">Option 2</Option>
      </CustomSelect>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });

  test("matches search enabled", () => {
    const result = render(
      <CustomSelect {...shared} searchable>
        <Option value="opt1">Option 1</Option>
        <Option value="opt2">Option 2</Option>
      </CustomSelect>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });

  test("matches floating variant", () => {
    const result = render(
      <CustomSelect variant="floating" {...shared}>
        <Option value="opt1">Option 1</Option>
        <Option value="opt2">Option 2</Option>
      </CustomSelect>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });
});
