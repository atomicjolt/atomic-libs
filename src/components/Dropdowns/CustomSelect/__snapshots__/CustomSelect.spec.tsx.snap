// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshots > matches default variant 1`] = `
<DocumentFragment>
  <div
    class="sc-imWZod kWDnYb aje-select aje-select--default is-medium"
  >
    <div
      aria-hidden="true"
      data-a11y-ignore="aria-hidden-focus"
      data-react-aria-prevent-focus="true"
      data-testid="hidden-select-container"
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        style="font-size: 16px;"
        tabindex="-1"
        type="text"
      />
      <label>
        Custom dropdown label
        <select
          tabindex="-1"
        >
          <option />
          <option
            value="opt1"
          >
            Item 1
          </option>
          <option
            value="opt2"
          >
            Item 2
          </option>
        </select>
      </label>
    </div>
    <label
      class="sc-kpDprT jfpflK aje-label is-auto"
      id="react-aria-:r3:"
    >
      Custom dropdown label
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="react-aria-:r7: react-aria-:r3:"
      class="sc-iGgVNO sc-gsFSjX gITxVe aje-btn sc-kAycRU bPlNWz aje-btn--dropdown is-medium"
      id="react-aria-:r2:"
      role="link"
      tabindex="0"
      type="button"
    >
      <span
        class="sc-dhKdPU ebBQqz"
        id="react-aria-:r7:"
      >
        Select an option
      </span>
      <i
        aria-hidden="true"
        class="sc-eDPFhE cPIjzF material-icons is-medium"
      >
        arrow_drop_down
      </i>
    </button>
  </div>
</DocumentFragment>
`;

exports[`matches snapshots > matches floating variant 1`] = `
<DocumentFragment>
  <div
    class="sc-imWZod kWDnYb aje-select aje-select--floating is-medium"
  >
    <div
      aria-hidden="true"
      data-a11y-ignore="aria-hidden-focus"
      data-react-aria-prevent-focus="true"
      data-testid="hidden-select-container"
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        style="font-size: 16px;"
        tabindex="-1"
        type="text"
      />
      <label>
        Custom dropdown label
        <select
          tabindex="-1"
        >
          <option />
          <option
            value="opt1"
          >
            Item 1
          </option>
          <option
            value="opt2"
          >
            Item 2
          </option>
        </select>
      </label>
    </div>
    <div
      class="sc-fPXMhL khXgbG"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="react-aria-:rf: react-aria-:rb:"
        class="sc-iGgVNO sc-gsFSjX gITxVe aje-btn sc-kAycRU bPlNWz aje-btn--dropdown is-medium"
        id="react-aria-:ra:"
        role="link"
        tabindex="0"
        type="button"
      >
        <span
          class="sc-dhKdPU ebBQqz"
          id="react-aria-:rf:"
        />
        <i
          aria-hidden="true"
          class="sc-eDPFhE cPIjzF material-icons is-medium"
        >
          arrow_drop_down
        </i>
      </button>
      <label
        class="sc-kpDprT jfpflK aje-label is-auto"
        id="react-aria-:rb:"
      >
        Custom dropdown label
      </label>
    </div>
  </div>
</DocumentFragment>
`;
